### PR TITLE
Split km artifacts

### LINKED ISSUE #745: Come up with a better artifact layout
Libraries that process annotations and parse Kotlin metadata represent only a subset of KotlinPoet consumers, whereas other consumers, such as SQLDelight and Wire, have other types of input. Therefore, helpers for Metadata and Elements processing should be moved out of the main artifact, which should help keep it small and generic.

- Move recently introduced helpers for working with Kotlin metadata into a new artifact that depends on both `kotlinpoet` and `kotlinpoet-km`. **Milestone**: 1.4.
- Find good names for `kotlinpoet-km` and the new artifact. **Milestone**: 1.4.
- Move helpers for processing Elements into a new artifact. **Milestone**: 2.0.

### PR DESCRIPTION
**Pull request by ZacSweers**
Resolves #745 

This splits artifacts down the following way:
- `kotlinpoet-km-specs`: `KmSpecs` and `ElementHandler`
- `kotlinpoet-elementhandler-elements`: Elements-based implementation of `ElementHandler`
- `kotlinpoet-elementhandler-reflective`: Reflective-based implementation of `ElementHandler`
- `km-specs-test`: `KmSpecsTest` extracted so it can depend on the 3 above

Per offline discussion with @Egorand - keeping the tiny bit of duplication in elementhandlers for now.

### CODE REVIEW

**Comment by ZacSweers** [comment]
Up for naming/structure suggestions. I think these are roughly the four artifacts we want though

File: buildSrc/src/main/java/Dependencies.kt [hunk 1]
```diff
@@ -18,7 +18,7 @@ object versions {
   const val spotless = "3.22.0"
   const val ktlint = "0.31.0"
   const val mavenPublish = "0.8.0"
-  const val shadowPlugin = "5.0.0"
+  const val shadowPlugin = "5.1.0"
```
Comment (Egorand) [inline buildSrc/src/main/java/Dependencies.kt]: Is this an unrelated change?

File: km-specs-tests/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  testImplementation(project(":kotlinpoet-km"))
+  testImplementation(project(":kotlinpoet-km-specs"))
+  testImplementation(project(":kotlinpoet-elementhandler-elements"))
+  testImplementation(project(":kotlinpoet-elementhandler-reflective"))
+  testImplementation(deps.kotlin.junit)
+  testImplementation(deps.test.truth)
+  testImplementation(deps.test.compileTesting)
+}
+
+repositories {
```
Comment (Egorand) [inline km-specs-tests/build.gradle.kts]: Root `build.gradle` applies a `repositories` block to all subproject, so I believe re-applying it here is not necessary.

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt [hunk 3]
```diff
@@ -475,6 +475,10 @@ class FunSpec private constructor(
       body.addStatement(format, *args)
     }
 
+    fun clearBody() = apply {
```
Comment (Egorand) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt]: Separate PR?

File: buildSrc/src/main/java/Dependencies.kt (hunk 1 above)
Comment (Egorand) [reply buildSrc/src/main/java/Dependencies.kt]: Ah, found the shading config which probably requires a version bump.

File: km-specs-tests/build.gradle.kts (hunk 2 above)
Comment (ZacSweers) [reply km-specs-tests/build.gradle.kts]: c2c64bd

**Comment by ZacSweers** [comment]
> Looks good + a couple small things. Did shading work as expected?

Yup! Manually verified looking at the jar, also it's tested by virtue of the separate test project consuming it

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt (hunk 3 above)
Comment (ZacSweers) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt]: #752

**Comment by Egorand** [comment]
Spotless is not happy, ready to merge once it's fixed.

**Comment by ZacSweers** [comment]
fixed and rebased 👍
